datafusion/expr-common/src/signature.rs

```diff
@@ -486,6 +486,145 @@ impl TypeSignature {
         }
     }
 
+    /// Return string representation of the function signature with parameter names.
+    ///
+    /// This method is similar to [`Self::to_string_repr`] but uses parameter names
+    /// instead of types when available. This is useful for generating more helpful
+    /// error messages.
+    ///
+    /// # Arguments
+    /// * `parameter_names` - Optional slice of parameter names. When provided, these
+    ///   names will be used instead of type names in the output.
+    ///
+    /// # Examples
+    /// ```
+    /// # use datafusion_expr_common::signature::TypeSignature;
+    /// # use arrow::datatypes::DataType;
+    /// let sig = TypeSignature::Exact(vec![DataType::Int32, DataType::Utf8]);
+    ///
+    /// // Without names: shows types
+    /// assert_eq!(sig.to_string_repr_with_names(None), vec!["Int32, Utf8"]);
+    ///
+    /// // With names: shows parameter names
+    /// assert_eq!(
+    ///     sig.to_string_repr_with_names(Some(&["id".to_string(), "name".to_string()])),
+    ///     vec!["id, name"]
+    /// );
+    /// ```
+    pub fn to_string_repr_with_names(
+        &self,
+        parameter_names: Option<&[String]>,
+    ) -> Vec<String> {
+        match self {
+            TypeSignature::Exact(types) => {
+                if let Some(names) = parameter_names {
+                    vec![names
+                        .iter()
+                        .take(types.len())
```

Comment:
This `take` (and in other branches too) feels a little odd; it feels this is something we should already know (that the lengths match) and thus we don't need to constantly check it, or something we check up front, instead of implicitly assuming here. Or do I misunderstand the purpose?